README.md
# OTMql4Fix
## Open Trading Metaquotes4 FIX Bridge

### OTMql4Fix - Open Trading Metaquotes4 FIX Bridge
https://github.com/OpenTrading/OTMql4Fix/

This package allows the OpenTrading Metatrader-Python bridge
(https://github.com/OpenTrading/OTMql4Py/)
to communicate with outside applications that speak the FIX protocol.
FIX is emerging standard for formatting messages for asynchronous
communications between financial and trading applications:
http://fixprotocol.org

This project builds on the FIX parser:
http://github.com/dmitryme/fix_parser
and builds on OTMql4Py, and requires that as a pre-requisite.
There are no pre-requisites in your Python other than `ctypes`,
which is now a part of the Python standard library. The `fix_parser`
works under Windows and Linux, so you can use it from Mt4 under Windows,
and use the `fix_parser` again in the client programs that communicate to Mt4,
under Windows or Linux.

The current code offers only the ability to format and parse FIX messages:
you will still need a communications mechanism to asynchronously
send and receive these formatted messages, such as OTMql4Fix:
https://github.com/OpenTrading/OTMql4Fix/

**This is a work in progress - a developers' pre-release version.**

It works on Mt4 builds > 6xx, but the documentation still needs to be done,
as well as more tests and testing on different versions.
Only Python 2.7.x is supported.

### History

The original idea for this project was to make a Mt4<->FIX
bi-directional bridge based on an existing C-coded gateway:
Http/sourceforge.net/projects/fixmt4gateway/files/fixGateway-mt4.zip
But that code has many problems, and is an incomplete, hard-coded,
partial implementation of FIX.

A better, more complete, Fix parser that is multi-FIX-version is:
http://github.com/dmitryme/fix_parser
which allows FIX to be cleanly integrated into Python. The code is very
efficient, and uses Python's `ctypes import` mechanism to expose
C/C++ code directly into Python. We expect this code to be faster
than code based on QuickFix (http://www.quickfix.org) which uses SWIG,
but we have not yet benchmarked the code.

We have simply compiled the `fix_parser` code according to its instructions
and have included the necessary glue to make it work under OTMql4Py,
and it install it the way Mt4 requires.
There are no changes to the `fix_parser` code itself.

To recompile the binaries we used MSVS2010Express from Microsoft,
and `cmake` from http://www.cmake.org:
http://www.cmake.org/files/v3.2/cmake-3.2.2-win32-x86.zip
The `fix_parser.dll` uses the `libxml2` libraries from http://www.xmlsoft.org,
and we used the compiled DLL supplied with the `fix_parser` distribution.

The project wiki should be open for editing by anyone logged into GitHub:
**Please report any system it works or doesn't work on in the wiki:
include the Metatrader build number, the origin of the metatrader exe,
the Windows version, and the Python version and origin of the Python.**
This code in known to run under Linux Wine (1.7.x), so this project
bridges Metatrader to Python under Linux.

### Installation

For the moment there is no installer: just "git clone" or download the
zip from github.com and unzip into an empty directory. Then recursively copy
the folder MQL4 over the MQL4 folder of your Metatrader installation. It will
not overwrite any Mt4 system files.

You will then need to copy the required DLLs from `lib/windows-x86`
into your Windows system folder (e.g. `c:\windows\system32`).
You may get away with copying them to any directory that is in your
Windows `PATH` environment variable, but your mileage may vary.

### Testing

Tests are yet to be written, but look at the example.py file in the
`MQL4/Python/OTMql427` directory.


### Project

Please file any bugs in the issue tracker:
https://github.com/OpenTrading/OTMql4Fix/issues

Use the Wiki to start topics for discussion:
https://github.com/OpenTrading/OTMql4Fix/wiki
It's better to use the wiki for knowledge capture, and then we can pull
the important pages back into the documentation in the share/doc directory.
You will need to be signed into github.com to see or edit in the wiki.



CHANGELOG.txt
commit eb4a65beb3cec7ca05ac186061adaec093f5e332
Date:   Tue Apr 28 16:34:16 2015 +0300

    Included the compiled lib/linux-x86/libfix_parser.so
    in case people wanted to run clients under Linux.

commit 4e533ec6d449bb27215ca2663490495c48722846
Date:   Tue Apr 28 11:57:26 2015 +0300

    Checked in the Python code with an example.
    We'll do an example test Script or Expert Real Soon Now(TM).

commit eda23d6a8bfc7e6260ac1f186ff1819ee02f0b47
Date:   Tue Apr 28 10:50:40 2015 +0300

    Added the compiled dlls from fix_parser.

commit c6c0bbc6c00e934893ee6dab5b9ba84bbe566242
Date:   Thu Apr 16 20:30:40 2015 +0300

    First cimmit.

commit ebcd1356fac558d44e9c3ebf13286281ff23863e
Date:   Thu Apr 16 17:22:41 2015 +0000

    Initial commit


MQL4/Python/OTMql427/PyFixExample.py
# -*-mode: python; py-indent-offset: 4; indent-tabs-mode: nil; encoding: utf-8-dos; coding: utf-8 -*-

# This code is taken from
# https://github.com/dmitryme/fix_parser/

import sys, os
from fix_fields import *
from fix_parser import *

import OTMql427
pDir = os.path.dirname(OTMql427.__file__)
pFile = b"fix.4.4.xml"
if pDir:
    pDir = os.path.join(os.path.dirname(pDir), 'etc', 'fix_descr')
    assert os.path.isdir(pDir), \
        "Installation error: cannot find the directory " +pDir
    pFile = os.path.join(bytes(pDir), pFile)
    
p = FixParser(pFile, PARSER_FLAG_CHECK_ALL)
del pDir, pFile, OTMql427

print("PROT VERSION = {0}".format(p.getProtocolVer()))
# create Exec Report
m = p.createMsg("8")
# check its params
print("MSG TYPE = {0}".format(m.getType()))
print("MSG NAME = {0}".format(m.getName()))

# fill message
m.setFieldAsString(FIXFieldTag_SenderCompID, b"QWERTY_12345678")
m.setFieldAsString(FIXFieldTag_TargetCompID, b"ABCQWE_XYZ")
m.setFieldAsInt32(FIXFieldTag_MsgSeqNum, 34)
m.setFieldAsString(FIXFieldTag_TargetSubID, "bsrv-ivanov_ii1")
m.setFieldAsString(FIXFieldTag_SendingTime, b"20120716-06:00:16.230")
m.setFieldAsString(FIXFieldTag_OrderID, b"1")
m.setFieldAsString(FIXFieldTag_ClOrdID, b"CL_ORD_ID_1234567")
m.setFieldAsString(FIXFieldTag_ExecID, b"FE_1_9494_1")
m.setFieldAsChar(FIXFieldTag_ExecType, b"0")
m.setFieldAsChar(FIXFieldTag_OrdStatus, b"1")
m.setFieldAsString(FIXFieldTag_Account, b"ZUM")
m.setFieldAsString(FIXFieldTag_Symbol, b"RTS-12.12")
m.setFieldAsChar(FIXFieldTag_Side, b"1")
m.setFieldAsDouble(FIXFieldTag_OrderQty, 25.0)
m.setFieldAsDouble(FIXFieldTag_Price, 135155.0)
m.setFieldAsChar(FIXFieldTag_TimeInForce, b"0")
m.setFieldAsDouble(FIXFieldTag_LastQty, 0)
m.setFieldAsDouble(FIXFieldTag_LastPx, 0)
m.setFieldAsDouble(FIXFieldTag_LeavesQty, 25.0)
m.setFieldAsDouble(FIXFieldTag_CumQty, 0)
m.setFieldAsDouble(FIXFieldTag_AvgPx, 0)
m.setFieldAsChar(FIXFieldTag_HandlInst, b"1")
m.setFieldAsString(FIXFieldTag_Text, b"COMMENT12")

# convert msg to string, \1 delimiter replaced with '|', just for pretty printing. '|' code is 124
str = m.toString(124)
print("MSG = {0}".format(str))

# parse input string and create FIX message
m1 = p.parse(str, 124)

# just print several fields of m1, to make sure str parsed ok
print("MSG1 TYPE = {0}".format(m1.getType()))
print("MSG1 NAME = {0}".format(m1.getName()))
print("SenderCompID = {0}".format(m1.getFieldAsString(FIXFieldTag_SenderCompID)))
print("TargetCompID = {0}".format(m1.getFieldAsString(FIXFieldTag_TargetCompID)))


MQL4/Python/OTMql427/fix_parser.py
# -*-mode: python; py-indent-offset: 4; indent-tabs-mode: nil; encoding: utf-8-dos; coding: utf-8 -*-

# This code is taken from
# https://github.com/dmitryme/fix_parser/

import platform
from ctypes import *
from types import *

# Assume that this is being called from the Python directory

import sys, os
from ctypes.util import find_library
if platform.system() == 'Windows':
   pFile = 'fix_parser'
   pFound = find_library('fix_parser')
   assert pFound and os.path.isfile(pFile), \
      "Python cannot find your " +pFile +".dll - put it in c:\\windows\\system32"
   try:
      lib = cdll.LoadLibrary(pFile)
   except Exception, e:
      # WindowsError: [Error 126] Module not found
      sys.stderr.write("Python probably cannot load libxml2.dll needed by " +pFile +".dll - make sure libxml2.dll is in c:\\windows\\system32"+"\n")
      raise
else:
   pFile = 'libfix_parser.so'
   pFound = find_library('fix_parser')
   assert pFound and os.path.isfile(pFound), \
      "Python cannot find your " +pFile +" - put in a dir on LD_LIBRARY_PATH"
   lib = cdll.LoadLibrary(pFile)
   
del os, pFile, pFound

FIX_SOH = 1

FIX_SUCCESS                        = 0
FIX_NO_FIELD                       = 1

FIX_FAILED                         = -1
FIX_ERROR_FIELD_HAS_WRONG_TYPE     = -2
FIX_ERROR_FIELD_NOT_FOUND          = -3
FIX_ERROR_FIELD_TYPE_EXISTS        = -4
FIX_ERROR_GROUP_WRONG_INDEX        = -5
FIX_ERROR_XML_ATTR_NOT_FOUND       = -6
FIX_ERROR_XML_ATTR_WRONG_VALUE     = -7
FIX_ERROR_PROTOCOL_XML_LOAD_FAILED = -8
FIX_ERROR_UNKNOWN_FIELD            = -9
FIX_ERROR_WRONG_PROTOCOL_VER       = -10
FIX_ERROR_DUPLICATE_FIELD_DESCR    = -11
FIX_ERROR_UNKNOWN_MSG              = -12
FIX_ERROR_LIBXML                   = -13
FIX_ERROR_INVALID_ARGUMENT         = -14
FIX_ERROR_MALLOC                   = -15
FIX_ERROR_UNKNOWN_PROTOCOL_DESCR   = -16
FIX_ERROR_NO_MORE_PAGES            = -17
FIX_ERROR_NO_MORE_GROUPS           = -18
FIX_ERROR_TOO_BIG_PAGE             = -19
FIX_ERROR_NO_MORE_SPACE            = -20
FIX_ERROR_PARSE_MSG                = -21
FIX_ERROR_WRONG_FIELD              = -22
FIX_ERROR_INTEGRITY_CHECK          = -23
FIX_ERROR_NO_MORE_DATA             = -24
FIX_ERROR_WRONG_FIELD_VALUE        = -25

PARSER_FLAG_CHECK_CRC             = 0x01       # check FIX message CRC during parsing
PARSER_FLAG_CHECK_REQUIRED        = 0x02       # check for required FIX fields
PARSER_FLAG_CHECK_VALUE           = 0x04       # check for valid value.
PARSER_FLAG_CHECK_UNKNOWN_FIELDS  = 0x08       # heck for unknown FIX fields during parsing. If not set all unknown fields ignored
PARSER_FLAG_CHECK_ALL             = PARSER_FLAG_CHECK_CRC | PARSER_FLAG_CHECK_REQUIRED | PARSER_FLAG_CHECK_VALUE | PARSER_FLAG_CHECK_UNKNOWN_FIELDS # make all possible checks

def raise_error(err):
   c_err_code = lib.fix_error_get_code(err)
   c_err_txt = lib.fix_error_get_text(err)
   err_txt = string_at(c_err_txt)
   lib.fix_error_free(err)
   raise Exception(c_err_code, err_txt)

class FixParser(object):
   def __init__(self, protFile, flags):
      error = c_long(0)
      self.parser = lib.fix_parser_create(protFile, 0, flags, pointer(error))
      if not self.parser:
         raise_error(error)

   def __exit__(self, type, value, traceback):
      lib.fix_parser_free(self.parser)

   def getProtocolVer(self):
      ver = lib.fix_parser_get_protocol_ver(self.parser)
      return string_at(ver)

   def createMsg(self, type):
      error = c_long(0)
      msg = lib.fix_msg_create(self.parser, type, pointer(error))
      if not msg:
         raise_error(error)
      return FixMessage(msg)

   def parse(self, data, delimiter):
      error = c_long(0)
      stop = c_char()
      msg = lib.fix_parser_str_to_msg(self.parser, c_char_p(data), c_uint(len(data)), c_byte(delimiter),
            pointer(stop), pointer(error))
      if not msg:
         raise_error(error)
      return FixMessage(msg)

class FixGroup(object):
   msg = 0
   group = 0
   def __init__(self, msg, group):
      self.msg = msg
      self.group = group

   def setFieldAsString(self, fieldTag, value):
      error = c_long(0)
      ret = lib.fix_msg_set_string(self.msg, self.group, fieldTag, value, pointer(error))
      if ret:
         raise_error(error)

   def setFieldAsInt32(self, fieldTag, value):
      error = c_long(0)
      ret = lib.fix_msg_set_int32(self.msg, self.group, fieldTag, value, pointer(error))
      if ret:
         raise_error(error)

   def setFieldAsInt64(self, fieldTag, value):
      error = c_long(0)
      ret = lib.fix_msg_set_int64(self.msg, self.group, fieldTag, value, pointer(error))
      if ret:
         raise_error(error)

   def setFieldAsChar(self, fieldTag, value):
      error = c_long(0)
      ret = lib.fix_msg_set_char(self.msg, self.group, fieldTag, c_char(value[0]), pointer(error))
      if ret:
         raise_error(error)

   def setFieldAsDouble(self, fieldTag, value):
      error = c_long(0)
      ret = lib.fix_msg_set_double(self.msg, self.group, fieldTag, c_double(value), pointer(error))
      if ret:
         raise_error(error)

   def getFieldAsInt32(self, fieldTag):
      error = c_long(0)
      val = c_int32(0)
      ret = lib.fix_msg_get_int32(self.msg, self.group, fieldTag, pointer(val), pointer(error))
      if ret:
         raise_error(error)
      return val.value

   def getFieldAsInt64(self, fieldTag):
      error = c_long(0)
      ptr = pointer(error)
      val = c_int64(0)
      ret = lib.fix_msg_get_int64(self.msg, self.group, fieldTag, pointer(val), pointer(error))
      if ret:
         raise_error(error)
      return val.value

   def getFieldAsDouble(self, fieldTag):
      error = c_long(0)
      val = c_double(0.0)
      ret = lib.fix_msg_get_double(self.msg, self.group, fieldTag, pointer(val), pointer(error))
      if ret:
         raise_error(error)
      return val.value

   def getFieldAsChar(self, fieldTag):
      error = c_long(0)
      val = c_char()
      ret = lib.fix_msg_get_char(self.msg, self.group, fieldTag, pointer(val), pointer(error))
      if ret:
         raise_error(error)
      return val.value

   def getFieldAsString(self, fieldTag):
      error = c_long(0)
      val = c_int(0)
      length = c_int(0)
      ret = lib.fix_msg_get_string(
            self.msg, self.group, fieldTag, pointer(val), pointer(length), pointer(error))
      if ret:
         raise_error(error)
      return string_at(val.value, length.value)

   def addGroup(self, fieldTag):
      error = c_long(0)
      group = lib.fix_msg_add_group(self.msg, self.group, fieldTag, pointer(error))
      if not group:
         raise_error(error)
      return FixGroup(self.msg, group)

   def delGroup(self, fieldTag, groupIdx):
      error = c_long(0)
      ret = lib.fix_msg_del_group(self.msg, self.group, fieldTag, groupIdx, pointer(error))
      if ret:
         raise_error(error)

   def getGroup(self, fieldTag, idx):
      error = c_long(0)
      group = lib.fix_msg_get_group(self.msg, self.group, fieldTag, idx, pointer(error))
      if not group:
         raise_error(error)
      return FixGroup(self.msg, group)

   def delField(self, fieldTag):
      error = c_long(0)
      ret = lib.fix_msg_del_field(self.msg, self.group, fieldTag, pointer(error))
      if ret:
         raise_error(error)

class FixMessage(FixGroup):
   def __init__(self, msg):
      self.msg = msg
      FixGroup(msg, 0)

   def __exit__(self, type, value, traceback):
      lib.fix_msg_free(self.obj)

   def getType(self):
      return string_at(lib.fix_msg_get_type(self.msg))

   def getName(self):
      return string_at(lib.fix_msg_get_name(self.msg))

   def toString(self, delimiter):
      error = c_long(0)
      buf = create_string_buffer(1024)
      reqLen = c_uint(0)
      ret = lib.fix_msg_to_str(self.msg, c_byte(delimiter), buf, len(buf), pointer(reqLen), pointer(error))
      if ret:
         if not error and reqLen.value > len(buf):
            buf = create_string_buffer(reqLen.value)
            ret = lib.fix_msg_to_str(self.msg, c_byte(delimiter), buf, len(buf), pointer(reqLen), pointer(error))
            if ret:
               raise_error(error)
         else:
            raise_error(error)
      return buf.value
